### PR TITLE
Fix reading FastAPI request body twice. 

### PR DESCRIPTION
**Pull request by antonpirker**
Starlette/FastAPI is internally caching the request body if read via `request.json()` or `request.body()` but NOT when using `request.form()`. This leads to a problem when our Sentry Starlette integration wants to read the body data and also the users code wants to read the same data. 

Solution:
Force caching of request body for `.form()` calls too, to prevent error when body is read twice.

The tests where mocking `.stream()` and thus hiding this problem. So the tests have been refactored to mock the underlying `._receive()` function instead.

### CODE REVIEW

File: sentry_sdk/integrations/starlette.py [hunk 1]
```diff
@@ -516,12 +516,16 @@ async def raw_data(self):
...
-        curl -X POST http://localhost:8000/upload/somethign  -F username=Julian -F password=hello123
+        curl -X POST http://localhost:8000/upload/something -H "Content-Type: application/x-www-form-urlencoded" -d "username=kevin&password=welcome123"
+        curl -X POST http://localhost:8000/upload/something  -F username=Julian -F password=hello123
         """
         if multipart is None:
             return None
 
+        # Parse the body first to get it cached, as Starlette does not cache form() as it
+        # does with body() and json() https://github.com/encode/starlette/discussions/1933
+        await self.request.body()
```
Comment (sl0thentr0py) [inline sentry_sdk/integrations/starlette.py:527]: this is getting more and more hacky tbh.

I'd rather prefer we rewrite the whole `extract_request_info` thing to something that makes sense for starlette. The current structure was just copied from flask and might not make sense for another framework. The relevant goal is to extract as much helpful request metadata as possible, the actual logic depends on the framework and what it supports.

**Review by sl0thentr0py — CHANGES_REQUESTED** [review]

File: sentry_sdk/integrations/starlette.py (hunk 1 above)
Comment (antonpirker) [reply sentry_sdk/integrations/starlette.py:527]: I know this looks hacky. But for our integration to get the form data from the request calling .body() and then .form() is the only way to not break our users Starlette projects.

Comment (antonpirker) [reply sentry_sdk/integrations/starlette.py:537]: This reading of the `body()` is still there, but I have now refactored `extract_request_info` to make it clearer what is happening.

Comment (antonpirker) [reply sentry_sdk/integrations/starlette.py:537]: I guess this is now better understandable and enough "non hacky". What do you think @sl0thentr0py ?

File: tests/integrations/starlette/test_starlette.py [hunk 2]
```diff
@@ -243,82 +247,77 @@ async def test_starlettrequestextractor_cookies(sentry_init):
...
+    side_effect = [_mock_receive(msg) for msg in FORM_RECEIVE_MESSAGES]
+    starlette_request._receive = mock.Mock(side_effect=side_effect)
+
+    extractor = StarletteRequestExtractor(starlette_request)
+
+    form_data = await extractor.form()
+    assert form_data.keys() == PARSED_FORM.keys()
+    assert form_data["username"] == PARSED_FORM["username"]
+    assert form_data["password"] == PARSED_FORM["password"]
+    assert form_data["photo"].filename == PARSED_FORM["photo"].filename
```
Comment (hasier) [inline tests/integrations/starlette/test_starlette.py:282]: Maybe call `await request.stream()` or `body()` right at the end to ensure that there are no regressions? Maybe it'd be relevant to do the same in the rest of tests as well, just to make sure in all cases we still have access to the whole request stream/body?

File: tests/integrations/starlette/test_starlette.py [hunk 3]
```diff
@@ -243,82 +247,77 @@ async def test_starlettrequestextractor_cookies(sentry_init):
...
-    with mock.patch(
-        "starlette.requests.Request.stream",
-        return_value=AsyncIterator(json.dumps(BODY_JSON)),
-    ):
-        starlette_request = starlette.requests.Request(SCOPE)
-        extractor = StarletteRequestExtractor(starlette_request)
+    await extractor.request.form()
 
-        assert await extractor.raw_data() == bytes(json.dumps(BODY_JSON), "utf-8")
+    with pytest.raises(RuntimeError):
```
Comment (sl0thentr0py) [inline tests/integrations/starlette/test_starlette.py:319]: I don't understand this test, shouldn't we test that our stuff doesn't throw even if we try to access both form and body?

Comment (hasier) [reply tests/integrations/starlette/test_starlette.py:319]: I think this was following [my suggestion](https://github.com/getsentry/sentry-python/issues/1675#issuecomment-1302200430). My idea was to have a test that crashes when you upgrade to a `starlette` version where they have fixed [the caching inconsistency](https://github.com/encode/starlette/discussions/1933), as that means you may need to double check your implementation and maybe simplify the code, right @antonpirker?

Comment (antonpirker) [reply tests/integrations/starlette/test_starlette.py:319]: Yes. This test calls directly the `form()` from the Starlette request and not the `form()` from the extractor, that "fixes" the caching with internally calling `body()`.

Comment (sl0thentr0py) [reply tests/integrations/starlette/test_starlette.py:319]: ahh gotcha, thx for the explanation both!

**Review by sl0thentr0py — APPROVED** [review]
alright lgtm

**Pull request merged**